File: c_solutions_21-30/Euler_28.c
#include "euler_util.h"


long spiral_total(int spiral_len)
{
    long total = 0;

    for (int i=1; i < spiral_len + 1; i+=2)
        total += (i * i);

    return total;
}        

long grid(int step, int start_axis, int end_axis)
{
    long total = step;

    for (int i=start_axis; step + i < end_axis; i+=8) {
        step += i;
        total += step;
    }

    return total;
}

int main(int argc, char *argv[])
{
    float start = timeit();

    int diagonal_pos = 3;

    long total = spiral_total(1001);
    int end_axis = 1001994;

    for (int i=10; i < 15; i+=2) {
        total += grid(diagonal_pos, i, end_axis);
        diagonal_pos += 2;
        end_axis += 2;
    }

    float stop = timeit();

    printf("Answer: %ld\n", total);
    printf("Time: %f\n", stop - start);

    return 0;
}

File: Python/py_solutions_21-30/Euler_28.py
# What is the sum of the numbers on the diagonals in a 1001 by 1001 spiral formed in the same way?

from __future__ import print_function

import timeit
from itertools import takewhile

try:
    range = xrange
except NameError:
    pass

start = timeit.default_timer()

def grid(step, axis):
    total = step
    for i in takewhile(lambda x: x + step < axis[-1], axis):
        step += i
        total += step                        
    return total

 
def euler_28():
    diag = 3
    total = sum(map(lambda x: x**2, range(1, 1002, 2)))
    for i in range(10, 15, 2):
        total += grid(diag, range(i, 1002002, 8))
        diag += 2
    return total


print("Answer: {}".format(euler_28()))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))
